src/platform/context/egl.c

```diff
@@ -119,6 +163,64 @@ has_extension(const struct ctx *context, const char *extension)
    return false;
 }
 
+WLC_PURE static bool
+has_extension(const struct ctx *context, const char *extension)
+{
+   assert(context && extension);
+   return is_extension_supported(context->extensions, extension);
+}
+
+WLC_PURE static bool
+has_device_extension(const struct ctx *context, const char *extension)
+{
+   assert(context && extension);
+   return is_extension_supported(context->device_extensions, extension);
+}
+
+EGLDeviceEXT
+get_egl_device()
+{
+   EGLint num_devices, i;
+   EGLDeviceEXT *devices = NULL;
+   EGLDeviceEXT device = EGL_NO_DEVICE_EXT;
+   EGLBoolean ret;
+
+   PFNEGLQUERYDEVICESEXTPROC eglQueryDevicesEXT = (void*)eglGetProcAddress("eglQueryDevicesEXT");
+   PFNEGLQUERYDEVICESTRINGEXTPROC eglQueryDeviceStringEXT = (void*)eglGetProcAddress("eglQueryDeviceStringEXT");
+
+   const char *extensions = eglQueryString(EGL_NO_DISPLAY, EGL_EXTENSIONS);
+   if (!is_extension_supported(extensions, "EGL_EXT_device_base") &&
+       (!is_extension_supported(extensions, "EGL_EXT_device_enumeration") ||
+        !is_extension_supported(extensions, "EGL_EXT_device_query")))
+      return NULL;
+
+   ret = eglQueryDevicesEXT(0, NULL, &num_devices);
+   if (!ret || num_devices < 1)
+      return NULL;
+
+   devices = calloc(sizeof(EGLDeviceEXT), num_devices);
```

Comment:
I think rather than allocating, we can have some sensible upper limit like 255 or something. I doubt there is ever that many devices.